src/org/labkey/test/stress/Simulation.java

```diff
@@ -0,0 +1,434 @@
+     * @param activity an activity definition
+     * @see Activity
+     */
+    private void runActivity(Activity activity) throws ExecutionException, InterruptedException
+    {
+        List<Future<?>> futures = new ArrayList<>();
+        for (Activity.RequestParams request : activity.getRequests())
+        {
+            if (!stopped.get())
+            {
+                futures.add(_activityExecutor.submit(() -> {
+                    makeRequest(request);
+                    return null;
+                }));
+            }
+        }
+
+        for (Future<?> future : futures)
+        {
+            future.get();
+        }
+    }
+
+    /**
+     * Make a single API request and submit the results to {@link #_resultCollector}
+     * @param requestParams parameters defining the API request
+     */
+    private void makeRequest(Activity.RequestParams requestParams) throws InterruptedException
+    {
+        ApiTestCommand command = new ApiTestCommand(requestParams);
+        Timer timer = new Timer();
+        int statusCode = 0;
+        try
+        {
+            CommandResponse response = command.execute(_connection);
+            statusCode = response.getStatusCode();
+        }
+        catch (CommandException e)
+        {
+            statusCode = e.getStatusCode();
+        }
+        catch (IOException e)
+        {
+            statusCode = -1;
+        }
+        finally
+        {
+            _resultCollector.submitResult(new RequestResult(requestParams, statusCode, timer, simulationMetadata));
+        }
+    }
+
+    /**
+     * Builder class to define the parameters of a simulation
+     * <ul>
+     *     <li>
+     *         {@link #_connectionFactory} - used to generate an API connection for the simulation to use
+     *     </li>
+     *     <li>
+     *         {@link #activityDefinitions} - {@link Activity} list that defines the simulation. These are deserialized from {@link ApiTestsDocument} XML files.
+     *     </li>
+     *     <li>
+     *         {@link #maxActivityThreads} - the size of thread pool to use for requests
+     *     </li>
+     *     <li>
+     *         {@link #delayBetweenActivities} - the number of milliseconds to pause between activities
+     *     </li>
+     *     <li>
+     *         {@link #runOnce} - this will cause the simulation to stop after running through the activities once
+     *     </li>
+     * </ul>
+     * {@link #startSimulation(Function)} is the primary entry point once all parameters have been set
+     */
+    public static class Definition
+    {
+        private final Supplier<Connection> _connectionFactory;
+
+        private List<Activity> activityDefinitions = Collections.emptyList();
+        private int maxActivityThreads = 6; // This seems to be the number of parallel requests browsers handle
+        private int delayBetweenActivities = 5_000;
+        private boolean runOnce = false;
```

Comment:
Would it be valuable to make any of these configurable through environment variables or TeamCity parameters and default to these values if nothing is set?

Reply:
I was thinking that these values would be part of the scenario definition. In other words, having a different delay or number of threads would be a different scenario.
Now, a particular scenario might want to add TeamCity knobs to adjust these values for that scenario but there wouldn't be any baked in to the helpers.